### PR TITLE
Shard the GatedDeltaNet sequence under ici_context_parallelism

### LINKED ISSUE #4932: GatedDeltaNet does not support ici_context_parallelism
Setting `ici_context_parallelism > 1` doesn't shard the sequence inside `Qwen3_5GatedDeltaNet`. Two separate things prevent it.

**1. Three pspecs pin the sequence axis to replicated.** In `models/qwen3.py`, `qkv_pspec`, `g_beta_pspec` and `qkvz_pspec` are built as `(KV_BATCH, None, KV_HEAD, None)`. The `None` in the sequence position tells XLA to gather the full sequence onto every device, so a `with_sharding_constraint` undoes the sharding the mesh asks for.

**2. The inter-chunk recurrence is a sequential scan.** `jax_chunk_gated_delta_rule` runs `lax.scan` over chunks. A sequential scan can't be split across devices, because each chunk needs the state the previous chunk produced.

## The recurrence is affine, so it can be parallelized

The GatedDeltaNet inter-chunk step is

```
h_new = exp(g_last) * h + k_g^T (u - w h)
      = (exp(g_last) I - k_g^T w) h + k_g^T u
      = A h + B
```

`A` and `B` don't depend on `h`. Affine maps compose associatively:

```
(A2, B2) o (A1, B1) = (A2 A1, A2 B1 + B2)
```

So each device composes its local chunks into one `(A, B)` pair with a local scan, the pairs compose across devices with a log-depth prefix scan over `ppermute`, and each device replays locally from its incoming state. `A` and `B` are 128x128 for every published Qwen 3.5 size, so the cross-device payload is small.

## Status

Filed as #4968.

- 1,048,576 tokens train end to end on 256 v5p at `ctx=256`, loss 12.923 to 10.583, 485.5 tokens/s/chip. The same configuration trains under Pathways with `enable_single_controller`.
- `tests/unit/gdn_cp_test.py` checks the composition against a sequential single-device reference on a forced 8-device CPU mesh, with a negative control. The whole path matched the stock sequential scan to 1.1e-08 in the earlier all-gather form.
- The local composition needs `jax.checkpoint` on the scan body. Without it the autodiff residuals reach 103 GB at sequence 262,144.
- An all-gather across devices is O(D) per layer, about 38 GB across 48 GatedDeltaNet layers at `ctx=256`. The prefix scan is log2(D) exchanges holding one pair.

The module is `models/gdn_cp.py`. The axis reaches `jax_chunk_gated_delta_rule` as a `cp_axis` keyword built from whichever context knob is set.

Running any of this needs `per_device_batch_size = 1/ici_context_parallelism`. See #4933.

cc @mmcsa

### PR DESCRIPTION
**Pull request by WandLZhang**
Fixes #4932.

Two pspecs pinned the GDN sequence axis to `None`, and the inter-chunk recurrence was a sequential `lax.scan`. Together they mean `ici_context_parallelism` can't shard the GDN sequence at all, so neither 256K nor 1M is reachable at any chip count.

The recurrence is affine in the state:

```
h' = (exp(g_last)·I − k_gᵀ·w)·h + k_gᵀ·u  =  A·h + B
```

Each device folds its local chunks into one `(A, B)` pair with a `lax.scan`, not `associative_scan`, which would materialize the running composition per chunk, 17 GB per device at a million tokens. The pairs then compose across devices with a Hillis-Steele prefix scan over `ppermute`: log2(D) exchanges holding one pair, O(1) in the device count. `A` and `B` are 128x128 at every published Qwen 3.5 size, so the payload doesn't grow with the model.

An all-gather composes the pairs correctly too, and that's what the first push of this PR did. It materializes `[D, B, H, K, K]` and `[D, B, H, K, V]` on every device, which is O(D) per layer: about 38 GB across the 48 GatedDeltaNet layers at `ctx=256`. That term is what made raising the context axis cost memory instead of saving it.

`qkvz_pspec` had the sequence axis hardcoded replicated, which told XLA to gather the full sequence onto every device: at `ctx=2`, seq 32,768, six live `bf16[2,32768,16,512]` buffers of 1.07 GB each. Found by diffing XLA buffer assignment between `ctx=1` and `ctx=2`.

**Measured** on Qwen3.5-27B, 256 v5p, `ctx=256`, `fsdp=1`, one 4,096-token sequence per device: **1,048,576 tokens**, loss 12.923 to 10.583. Also 524,288 on the same slice, and 131,072 on 64 chips.

## The launch flag this needs

Set `per_device_batch_size` to `1/ici_context_parallelism`. `num_target_devices` counts every mesh axis, so the local batch is `per_device_batch_size × ctx`, and the default grows it with the context axis. With the flag set, memory tracks tokens per device rather than the context degree. See #4933.

## Interaction with #4348

The two can't both be live. The fused kernel returns above the state composition, so `use_pallas` is forced false whenever a context axis is set, and sequence-sharded runs take the XLA scan. @bzantium verified that fallback is loss-preserving: 6e-6 relative on the output, 2e-5 on final state. The two are disjoint rather than stacked, so this goes against `main` and #4348 rebases on top.

## Size

`models/gdn_cp.py` is new, 113 lines. `models/qwen3.py` is +42/-4. Applies to a pristine checkout of `main` with no dependency on #4348 or on my other open PRs.

# Tests

`ici_context_parallelism=4` at sequence 32,768 and `ici_context_parallelism=8` at 262,144, both training with loss descending, on v5p and v6e. 1,048,576 tokens on 256 v5p. Memory is flat in `ctx` at a fixed local shard: about 30.5 GB on a 2,048-token shard whether `ctx` is 2 or 16, so the sizing extrapolates.

`tests/unit/gdn_cp_test.py` covers three things:

- `compose_local`, against the recurrence applied chunk by chunk in its original form. That reference never builds `A` or `B` and never calls `compose`, so it checks the affine reformulation rather than restating it. Reversing the chunk order is the negative control.
- `incoming_state` on a single-device mesh, where a shard has nothing before it.
- `incoming_state` against a sequential composition on a forced 8-device CPU mesh, over a 1-D context mesh, a 2-D fsdp x context mesh, and a tuple axis name. Breaking the chain is the negative control.

The multi-device case runs as a subprocess, so the forced device count takes effect before JAX initializes, following `tests/unit/usp_collective_test.py`. pyink clean at `--line-length=122`.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):

- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

---

_2026-09-15: the 485.5 tok/s/chip this originally published is withdrawn — a two-step artifact of the step timer. Unprofiled over six steps it is 90.2 at `ctx=256` and 104.9 at `ctx=64`. The runs stand._

### CODE REVIEW

File: src/maxtext/kernels/attention/gdn_cp.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  A_all = lax.all_gather(A_loc, cp_axis, axis=0, tiled=False)
+  B_all = lax.all_gather(B_loc, cp_axis, axis=0, tiled=False)
+  A_cum, B_cum = lax.associative_scan(compose, (A_all, B_all), axis=0)
+
+  idx = lax.axis_index(cp_axis)
+  prev = jnp.maximum(idx - 1, 0)
+  carried = jnp.matmul(A_cum[prev], h_init, precision=_PREC) + B_cum[prev]
+  h_in = jnp.where(idx == 0, h_init, carried)
+  final_h = jnp.matmul(A_cum[-1], h_init, precision=_PREC) + B_cum[-1]
+  return h_in, final_h
```
[the remark this answers is not in the document]
Comment (WandLZhang) [reply src/maxtext/kernels/attention/gdn_cp.py:130]: `lax.axis_index` takes a tuple. On 8 CPU devices, `axis_index(("context",))` and `axis_index(("a","b"))` both return a flat index across the product, `all_gather` reports D=8 for both, and `ppermute` with a tuple axis and flat indices shifts as expected. The 1,048,576-token run used a 1-tuple on 256 chips.

`tests/unit/gdn_cp_test.py` covers it. The third case builds a 2x4 `("context", "context_usp_ulysses")` mesh and passes the tuple through unchanged.

On the suggestion itself: it restores the all-gather this PR replaces. Gathering the pairs materializes `[D, B, H, K, K]` and `[D, B, H, K, V]` on every device, O(D) per layer, about 38 GB across the 48 GatedDeltaNet layers at `ctx=256`. The prefix scan is log2(D) exchanges holding one pair.

**Review by igorts-git — APPROVED** [review]

**Review by darisoy — APPROVED** [review]

**Comment by darisoy** [comment]
Thanks for the PR!

Regarding file organization: would it make sense to move `src/maxtext/models/gdn_cp.py` to `src/maxtext/kernels/attention/` (e.g., `src/maxtext/kernels/attention/gdn_cp.py` or `src/maxtext/kernels/attention/gated_delta_network_cp.py`)?

`src/maxtext/models/` is generally reserved for top-level model family architectures (`qwen3.py`, `deepseek.py`, etc.), whereas attention/recurrence kernels and context parallelism collective utilities live under `src/maxtext/kernels/attention/` (where `gated_delta_network.py`, `tokamax_ring_attention.py`, `ulysses_attention.py`, and `usp_attention.py` reside).

File: src/maxtext/kernels/attention/gdn_cp.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
+"""Context-parallel evaluation of the GatedDeltaNet inter-chunk recurrence.
```
Comment (NuojCheng) [inline src/maxtext/kernels/attention/gdn_cp.py:14]: `context_parallel_load_balance` defaults as true with logic happened in https://github.com/AI-Hypercomputer/maxtext/blob/ab9f7c50c4934f75d893b295a4e9e8b3012e11da/src/maxtext/utils/train_utils.py#L308. However, context parallelism for linear attention does not need load balance. We need a form of guard to disable load balance otherwise the correctness fails.

File: tests/unit/gdn_cp_test.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
+# Copyright 2026 Google LLC
```
Comment (NuojCheng) [inline tests/unit/gdn_cp_test.py:1]: please add an integration test for hybrid model using context parallelism

File: src/maxtext/models/gdn_cp.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  # composition costs O(D) per device per layer. At ctx=256 on Qwen3.5-27B that
+  # is roughly 38 GB across the 48 GatedDeltaNet layers, and it is what makes
+  # per-device memory grow with the context axis instead of shrinking.
+  #
+  # Hillis-Steele instead: log2(D) ppermute exchanges, one (A, B) pair live at a
+  # time, O(1) in D, and log2(D) backward residuals rather than D.
+  #
+  # D comes from the static shape of an all-gathered scalar. That is 4D bytes and
+  # gives a Python int usable as a loop bound.
+  D = int(lax.all_gather(jnp.int32(1), cp_axis).shape[0])
```
Comment (NuojCheng) [inline src/maxtext/models/gdn_cp.py]: please use `jax.lax.axis_size` instead of AG to compute CP rank.

File: src/maxtext/models/gdn_cp.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
+"""Context-parallel evaluation of the GatedDeltaNet inter-chunk recurrence.
+
+The recurrence h_new = A @ h + B is affine in the state, so it composes
+associatively and can be split across a sharded sequence. See
+apply_gdn_context_parallel.py for the derivation and the measurements.
```
Comment (NuojCheng) [inline src/maxtext/models/gdn_cp.py]: what is `apply_gdn_context_parallel.py`?

File: src/maxtext/models/qwen3.py [hunk 6]
```diff
@@ -881,8 +899,17 @@ def extract_state(c_in, v_len):
...
-      g_beta_pspec = logical_to_mesh_axes((KV_BATCH, None, KV_HEAD), mesh=self.mesh, rules=logical_rules)
+      # LENGTH, not None. The sequence axis was hardcoded to replicated, so
+      # ici_context_parallelism could never shard the GDN sequence while still
+      # consuming the context axis from the mesh -- which is why raising ctx
+      # made memory worse instead of better. The scan handles a sharded
+      # sequence via the two-pass affine composition in models/gdn_cp.py.
+      # Either context axis can carry the sequence. LENGTH maps to both in the
+      # logical rules, so this is correct whichever one is configured.
+      _cp_on = cfg.ici_context_parallelism > 1 or getattr(cfg, "ici_context_usp_ulysses_parallelism", 1) > 1
+      _cp_len = LENGTH if _cp_on else None
```
Comment (NuojCheng) [inline src/maxtext/models/qwen3.py]: repeat with line 636-637, can we combine?

**Review by NuojCheng — COMMENTED** [review]
The main concern is how linear attention context parallelism works with load balance for a hybrid model architecture. We might need a separate PR fixing it otherwise we can only do ulysses-style CP for now. There could be some correctness issue using current PR. 

Besides that, some minor suggestions about performance and code layout. Thank you for the contribution!

**Comment by WandLZhang** [comment]
Thanks! the load balance point was a real bug. 

configs/types.py now checks qwen3_next and qwen3_5 at ctx>1 with load balance on

@darisoy: gdn_cp.py moved to kernels/attention/; 
lax.axis_size replaces the all-gather;
one gdn_context_axes() helper replaces three copies; 
the apply_gdn_context_parallel.py reference is gone

Integration test added: it runs jax_chunk_gated_delta_rule sharded against unsharded on an 8-device CPU mesh, comparing every output token and the final state. I also added gradient checks. there’s no custom_vjp here, so the backward is JAX's transpose rules composing

File: src/maxtext/kernels/attention/gdn_cp.py (hunk 2 above)
Comment (WandLZhang) [reply src/maxtext/kernels/attention/gdn_cp.py:14]: fixed in d5679bf. configs/types.py checks qwen3_next and qwen3_5 when ctx>1 and load balance is on

File: tests/unit/gdn_cp_test.py (hunk 3 above)
Comment (WandLZhang) [reply tests/unit/gdn_cp_test.py:1]: added in d5679bf. _run_end_to_end_checks runs jax_chunk_gated_delta_rule sharded against unsharded on an 8-device CPU mesh, comparing every output token and the final state

also added gradient checks

File: src/maxtext/models/gdn_cp.py (hunk 4 above)
Comment (WandLZhang) [reply src/maxtext/models/gdn_cp.py]: done in d5679bf. D = lax.axis_size(cp_axis), works with the tuple axis

File: src/maxtext/models/gdn_cp.py (hunk 5 above)
Comment (WandLZhang) [reply src/maxtext/models/gdn_cp.py]: it was in the patch series this PR came from and is not in this repo. removed in d5679bf, the docstring has the derivation now

File: src/maxtext/models/qwen3.py (hunk 6 above)
Comment (WandLZhang) [reply src/maxtext/models/qwen3.py]: done in d5679bf, one gdn_context_axes(cfg) helper
